[Account Settings] Add Ability to Change Username

danielebogo opened the pull request:
Refs. #9705

![username-change-2](https://user-images.githubusercontent.com/912252/62628520-5f936580-b923-11e9-93d0-b8c5db9039e6.gif)

This PR adds the ability to change the username. The feature is based on the _Flux + View Model_ pattern. The view controller is a subclass of `SignupUsernameTableViewController` in order to keep the same UI structure.

## To test:
- Select the Tab _Me_ -> _Account Settings_
- Tap on the _username_ to open the new screen and change the username
- If the user is an A8C user you shouldn't be able allowed to change the username

Update release notes:

- [x] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

frosty commented on WordPress/Classes/ViewRelated/Me/Change Username/ChangeUsernameViewController.swift:
```diff
@@ -0,0 +1,153 @@
+private extension ChangeUsernameViewController {
+    func setupViewModel() {
+        viewModel.reachabilityListener = { [weak self] in
+            DispatchQueue.main.async {
```
Should we just ensure that each of these listener blocks is called on the main queue from the viewModel?

danielebogo commented on WordPress/Classes/Stores/AccountSettingsStore.swift:
```diff
@@ -0,0 +1,134 @@
+import WordPressFlux
+
+enum AccountSettingsState: Equatable {
+    case stationary
```
Oh I like `idle`!!! Thanks a lot!

danielebogo commented:
Thanks @frosty for reviewing it. I update everything. I also added the local validation considering the alphanumeric charset.

I added the alert to confirm the save action. I'm not sure about the copy though.

CC @SylvesterWilmott 

![Simulator Screen Shot - iPhone X - 2019-08-01 at 16 25 16](https://user-images.githubusercontent.com/912252/62310463-819f6a80-b481-11e9-807d-eaae4ae03011.png)

![Simulator Screen Shot - iPhone X - 2019-08-01 at 17 20 29](https://user-images.githubusercontent.com/912252/62310469-8532f180-b481-11e9-8830-55c6f7112b52.png)

frosty commented on WordPress/Classes/ViewRelated/Me/Change Username/View Model/ChangeUsernameViewModel.swift:
```diff
@@ -132,6 +143,13 @@ private extension ChangeUsernameViewModel {
+
+private extension String {
+    var isAlphanumeric: Bool {
+        return range(of: "[^a-zA-Z0-9]", options: .regularExpression) == nil
```
Are usernames allowed accented characters at all? (I don't know the answer to this) I suspect not, but if they are we may want to do something like

`rangeOfCharacter(from: CharacterSet.alphanumerics.inverted) == nil`

danielebogo replied:
>rangeOfCharacter(from: CharacterSet.alphanumerics.inverted) == nil

This was my first attempt. Using characters sets to check the string, but it includes the accents. The regex then was the simple solution to get only alphanumerics characters ignoring diacritics.

danielebogo commented:
@frosty @SylvesterWilmott Calypso uses to add a confirmation textfield as soon you start typing in the _Username_ field. Although I think this is still the best experience, because with an alert the user can accidentally confirm the action, a second textfield (plus the actions row we can add in future) pushes the text to the bottom below the keyboard.
I was thinking if at this point we can use a button at the bottom of the page rather than the save button in the navigation bar. This forces the user to scroll the page and read the paragraph.
What do you think?

folletto commented:
Any reason why we aren't reusing the name suggestions from the signup flow to avoid people getting frustrated in guessing an available username?

danielebogo commented:
@folletto  we we are using the same experience of Calypso. There's an API call to validate what you type but not to suggest a new username. Should we stay with it or make a different flow? @SylvesterWilmott @frosty

folletto commented:
I haven't checked in a while so I'm not sure, but we already have a screen to do that in the signup flow! Already designed, with suggestions. 

See #8711 

To me _this_ is introducing a different flow, not the other way around, as the users are on mobile, and have already (maybe) experienced the other UI.

frosty commented on WordPress/Classes/ViewRelated/Me/Change Username/ChangeUsernameViewController.swift:
```diff
@@ -0,0 +1,162 @@
+        enum Alert {
+            static let loading = NSLocalizedString("Loading usernames", comment: "Shown while the app waits for the username suggestions web service to return during the site creation process.")
+            static let title = NSLocalizedString("Careful!", comment: "Alert title.")
+            static let message = NSLocalizedString("You are changing your Username to %@. Changing your username will also affect your Gravatar profile and IntenseDebate profile addresses. \nConfirm your new Username to continue.", comment: "Alert message.")
```
Nitpick, but a couple of occurrences of 'username' in here have a capital U. I think they should all be lowercase?

frosty commented on WordPress/Classes/ViewRelated/Me/Change Username/View Model/ChangeUsernameViewModel.swift:
```diff
@@ -0,0 +1,140 @@
+import WordPressFlux
+
+class ChangeUsernameViewModel {
+    typealias VoidListener = () -> Void
```
Perhaps just `Listener`?

frosty commented on WordPress/Classes/Services/AccountSettingsService.swift:
```diff
@@ -18,6 +18,7 @@ protocol AccountSettingsRemoteInterface {
     func getSettings(success: @escaping (AccountSettings) -> Void, failure: @escaping (Error) -> Void)
     func updateSetting(_ change: AccountSettingsChange, success: @escaping () -> Void, failure: @escaping (Error) -> Void)
     func changeUsername(to username: String, success: @escaping () -> Void, failure: @escaping () -> Void)
+    func validateUsername(to username: String, success: @escaping () -> Void, failure: @escaping (Error) -> Void)
```
Are we still using the validate methods now we changed to the suggestion model? If not, should we remove them?

frosty commented:
> Should we disable the Save button until a different (new) username is selected?

I can't recreate this at the moment, so I'm not quite sure how I got into that state 🤔

One other thing I forgot to mention: on the confirmation alert, if you type the new username incorrectly we just dismiss the alert and fail silently. We should probably let the user know that their username wasn't changed.

danielebogo commented:
Hey @frosty thanks for your review! I pushed some change and it's ready for another round!
Here you can find some screenshot about the alert with the button disabled/enabled:

![Simulator Screen Shot - iPhone X - 2019-08-08 at 15 47 02](https://user-images.githubusercontent.com/912252/62714061-bc5f5080-b9f5-11e9-997b-79e56caed54d.png)

![Simulator Screen Shot - iPhone X - 2019-08-08 at 15 47 06](https://user-images.githubusercontent.com/912252/62714063-bc5f5080-b9f5-11e9-8c87-cea032540724.png)

danielebogo commented:
Thanks @frosty ! It's ready again 😊
